### PR TITLE
Short circuited to MatchNone for non-participating  slice

### PR DESCRIPTION
**Pull request by nirmalc**
When the query is  expensive  parallel usage of scroll/slice API overloads cluster, ( In my case a `span_multi` query causes heap / gc issues )

Looks like its because same query is processed against all non-matching shards with a `MatchNoDocs` and matching-shards with `MatchAllDocs` filter. In case of number of shards =< number of slices, returning a `MatchNoDocs` query instead of filter seems to avoid that problem.

### CODE REVIEW

**Comment by mayya-sharipova** [comment]
@elasticmachine  test this please

File: server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java [hunk 1]
```diff
@@ -274,7 +275,12 @@ public Query buildFilteredQuery(Query query) {
         }
 
         if (sliceBuilder != null) {
-            filters.add(sliceBuilder.toFilter(clusterService, request, queryShardContext));
+            Query slicedQuery = sliceBuilder.toFilter(clusterService, request, queryShardContext);
+            if ( slicedQuery instanceof MatchNoDocsQuery){
+                return slicedQuery;
+            }else {
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java]: nit: space before `else`

File: server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java [hunk 2]
```diff
@@ -274,7 +275,12 @@ public Query buildFilteredQuery(Query query) {
         }
 
         if (sliceBuilder != null) {
-            filters.add(sliceBuilder.toFilter(clusterService, request, queryShardContext));
+            Query slicedQuery = sliceBuilder.toFilter(clusterService, request, queryShardContext);
+            if ( slicedQuery instanceof MatchNoDocsQuery){
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java]: extra space after `(`

File: server/src/test/java/org/elasticsearch/search/DefaultSearchContextTests.java [hunk 3]
```diff
@@ -178,6 +182,20 @@ public void testPreProcess() throws Exception {
             ParsedQuery parsedQuery = ParsedQuery.parsedMatchAllQuery();
             context3.sliceBuilder(null).parsedQuery(parsedQuery).preProcess(false);
             assertEquals(context3.query(), context3.buildFilteredQuery(parsedQuery.query()));
+
+            when(queryShardContext.getIndexSettings()).thenReturn(indexSettings);
+            when(indexService.newQueryShardContext(anyInt(),anyObject(),anyObject(),anyString())).thenReturn(queryShardContext);
+            when(queryShardContext.fieldMapper(anyString())).thenReturn(mock(MappedFieldType.class));
+            when(shardSearchRequest.indexRoutings()).thenReturn(new String[0]);
+
+            DefaultSearchContext context4 = new DefaultSearchContext(3L, shardSearchRequest, target, searcher, null,
```
Comment (mayya-sharipova) [inline server/src/test/java/org/elasticsearch/search/DefaultSearchContextTests.java]: nit: `new DefaultSearchContext(3L,` =>  `new DefaultSearchContext(4L,`

File: server/src/test/java/org/elasticsearch/search/DefaultSearchContextTests.java [hunk 4]
```diff
@@ -178,6 +182,20 @@ public void testPreProcess() throws Exception {
             ParsedQuery parsedQuery = ParsedQuery.parsedMatchAllQuery();
             context3.sliceBuilder(null).parsedQuery(parsedQuery).preProcess(false);
             assertEquals(context3.query(), context3.buildFilteredQuery(parsedQuery.query()));
+
+            when(queryShardContext.getIndexSettings()).thenReturn(indexSettings);
+            when(indexService.newQueryShardContext(anyInt(),anyObject(),anyObject(),anyString())).thenReturn(queryShardContext);
```
Comment (mayya-sharipova) [inline server/src/test/java/org/elasticsearch/search/DefaultSearchContextTests.java]: Looks like, line 187 in not necessary

**Review by mayya-sharipova — APPROVED** [review]
thanks @nirmalc, nice enhancement!
I left small comments to be addressed

File: server/src/test/java/org/elasticsearch/search/DefaultSearchContextTests.java (hunk 4 above)
Comment (nirmalc) [reply server/src/test/java/org/elasticsearch/search/DefaultSearchContextTests.java]: Thanks for review , it is not necessary. I think i added in my first attempt to include use-case with routings in test. I've reformatted code with intelliJ   - sorry for formatting issues.

**Comment by mayya-sharipova** [comment]
@elasticmachine test this please

**Comment by mayya-sharipova** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Comment by mayya-sharipova** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by mayya-sharipova** [comment]
@elasticmachine update branch

**Comment by mayya-sharipova** [comment]
@elasticmachine test this please

**Comment by mayya-sharipova** [comment]
@nirmalc  Thank you for the PR, your change has ben merged, and will be available from 7.7

**Pull request merged**